language/move-lang/src/lib.rs

```diff
@@ -507,38 +607,74 @@ impl PassResult {
     }
 }
 
-fn run(cur: PassResult, until: Pass) -> Result<PassResult, Errors> {
+fn run(
+    pre_compiled_lib: Option<&FullyCompiledProgram>,
+    cur: PassResult,
+    until: Pass,
+    mut result_check: impl FnMut(&PassResult),
```

Comment:
I found this callback/hook quite hard to understand. I wonder if it would be better if we simply get rid of this function and let its clients take control of the steps:
```
let s0 = ...;
let s1 = phase1(s0);
let s2 = phase2(s1);
...
```

Reply:
If you force people to do that, you just force them to re-implement the main logic of the relationship between pass and pass result, and what the next step is. 

In other words, the call backs makes it so every use site that needs something like it doesn't have to reimplement the main switch statement for current state and the next one